### PR TITLE
[Imaging-216] Add support for Alpha channel in RGB images

### PR DESCRIPTION
**Pull request by gwlucastrig**
Here is my PR for Imaging-216.

### CODE REVIEW

File: src/test/java/org/apache/commons/imaging/formats/tiff/TiffReadAlphaTest.java [hunk 1]
```diff
@@ -0,0 +1,91 @@
...
+     * data.
+     *
+     * @param name a valid file name
+     * @return a valid file reference.
+     */
+    private File getTiffFile(String name) {
+        File tiffFolder = new File(ImagingTestConstants.TEST_IMAGE_FOLDER, "tiff");
+        File alphaFolder = new File(tiffFolder, "12");
+        return new File(alphaFolder, name);
+    }
```
Comment (kinow) [inline src/test/java/org/apache/commons/imaging/formats/tiff/TiffReadAlphaTest.java:64]: I think this is a windows CRLF?

File: src/test/java/org/apache/commons/imaging/formats/tiff/TiffReadAlphaTest.java [hunk 2]
```diff
@@ -0,0 +1,91 @@
...
+                    int p = testSite[i][2];
+                    int t = composite.getRGB(x, y);
+                    assertEquals(t, p, "Error for "+name+" at position "+x+", "+y);
+                }
+            }catch(ImageReadException | IOException ex){
+                fail("Exception reading "+name+", "+ex.getMessage());
+            }
+        }
+    }
+}
```
Comment (kinow) [inline src/test/java/org/apache/commons/imaging/formats/tiff/TiffReadAlphaTest.java:91]: Missing newline

File: src/test/java/org/apache/commons/imaging/formats/tiff/TiffAlphaRoundTripTest.java [hunk 3]
```diff
@@ -0,0 +1,137 @@
...
+/**
+ * Performs a round-trip that writes an image containing Alpha and then reads it back.
+ * Selected non-opaque pixels are tested for correctness,
+ */
+public class TiffAlphaRoundTripTest {
+
+
+    @TempDir
+    Path tempDir;
+
```
Comment (kinow) [inline src/test/java/org/apache/commons/imaging/formats/tiff/TiffAlphaRoundTripTest.java:45]: Strange character, probably Windows CRLF?

File: src/test/java/org/apache/commons/imaging/formats/tiff/TiffAlphaRoundTripTest.java [hunk 4]
```diff
@@ -0,0 +1,137 @@
...
+     * @return true if the components of the values match
+     */
+    boolean componentMatch(int a, int b, int iShift, int iTolerance){
+        int delta = ((a >> iShift) & 0xff) - ((b >> iShift) & 0xff);
+        if(delta<0){
+            delta = -delta;
+        }
+        return delta<iTolerance;
+    }
+}
```
Comment (kinow) [inline src/test/java/org/apache/commons/imaging/formats/tiff/TiffAlphaRoundTripTest.java:137]: Missing newline

File: src/test/java/org/apache/commons/imaging/common/ImageBuilderTest.java [hunk 5]
```diff
@@ -0,0 +1,146 @@
...
+
+        ImageBuilder testBuilder = imageBuilder.getSubset(25, 25, 25, 25);
+        for(int x=25; x<50; x++){
+            for(int y=25; y<50; y++){
+                int k = testBuilder.getRGB(x-25, y-25);
+                int rgb = imageBuilder.getRGB(x, y);
+                assertEquals(k, rgb, "Invalid image builder subpixel at "+x+", "+y);
+            }
+        }
+    }
```
Comment (kinow) [inline src/test/java/org/apache/commons/imaging/common/ImageBuilderTest.java:87]: These tests :point_up:  are for the other change with subset right? Just to confirm :+1:

File: src/main/java/org/apache/commons/imaging/formats/tiff/constants/TiffConstants.java [hunk 6]
```diff
@@ -70,6 +70,15 @@
...
 
+
+    /**
+     * Specifies that an application-specified photometric interpreter
+     * is to be used when reading TIFF files to convert raster data samples
+     * to RGB values for the output image.
+     * <p>
+     * The value supplied with this key should be a valid instance of
+     * a class that implements PhotometricInterpreter.
+     */
```
Comment (kinow) [inline src/main/java/org/apache/commons/imaging/formats/tiff/constants/TiffConstants.java:81]: :+1:

File: src/main/java/org/apache/commons/imaging/common/ImageBuilder.java [hunk 7]
```diff
@@ -200,27 +244,7 @@ public ImageBuilder getSubset(final int x, final int y, final int w, final int h
...
-        }
-        if (y < 0 || y >= height) {
-            throw new RasterFormatException("subimage y is outside raster");
-        }
-        if (y + h > height) {
-            throw new RasterFormatException(
-                    "subimage (y+height) is outside raster");
-        }
-
+        checkBounds(x, y, w, h);
```
Comment (kinow) [inline src/main/java/org/apache/commons/imaging/common/ImageBuilder.java:247]: :+1: Nice, avoiding code duplication!

**Review by kinow — COMMENTED** [review]
Reviewed 7 out of the 12 files (the tests and smaller changes). Some files will require understanding what's being changed, and some of the implementation logic, so will need more time.

Left comments for anything I found, but feel free to discard in case I missed something (didn't have much time to review this one now, sorry).

Thanks Gary!
Bruno

**Comment by gwlucastrig** [comment]
Not sure what's going on with the Windows newline issue (the ASCII 13 or ^M character).  Normally, git automatically fixes those.  I will go through those files and clean them up by hand.

Also, I'm think my bad luck (operator error) with git may still be in force.  Looking at the maven output reports, I find the following lines... Are they important?

Warning:  Cannot get the branch information from the git repository: 
Detecting the current branch failed: fatal: ref HEAD is not a symbolic ref

File: src/test/java/org/apache/commons/imaging/common/ImageBuilderTest.java (hunk 5 above)
Comment (gwlucastrig) [reply src/test/java/org/apache/commons/imaging/common/ImageBuilderTest.java:87]: Yes.  I'm trying to bring the coverage back to 100 percent for ImageBuilder

**Comment by gwlucastrig** [comment]
I added fixes for your review comments, but for some reason, I can seem to get today's commit to be added to the pull request.  I must have made some mistake setting it up in git

I'm thinking that the best thing to do would be to cancel the pull request and start over.

**Comment by kinow** [comment]
> Not sure what's going on with the Windows newline issue (the ASCII 13 or ^M character). Normally, git automatically fixes those. I will go through those files and clean them up by hand.

Thanks! It may happen depending on the editor you have I think.

> Also, I'm think my bad luck (operator error) with git may still be in force. Looking at the maven output reports, I find the following lines... Are they important?
>
> Warning: Cannot get the branch information from the git repository:
> Detecting the current branch failed: fatal: ref HEAD is not a symbolic ref

It looks very strange. I think I haven't seen this error, except when I changed my remotes, or deleted some branches by accident?

>I added fixes for your review comments, but for some reason, I can seem to get today's commit to be added to the pull request. I must have made some mistake setting it up in git
>
>I'm thinking that the best thing to do would be to cancel the pull request and start over.

The pull request has a remote repository and a branch associated. I suspect if you modify them (by deleting and re-forking a project, for example) GitHub UI might get confused. Not necessarily git I think, more the way GitHub is built and expects pull requests to be.

Might be easier to start over, just take care to save your changes somewhere first, so that you don't lose them :+1: (it never hurts to temporarily copy a project folder somewhere else while performing some git-fu)

**Comment by gwlucastrig** [comment]
Thanks.   I'm going to get rid of the current fork and start over.

I think that my next "off hours" task is going to have to be to spend some
time learning git for real (instead of just following canned procedures I
find on the web). I'm going to have to put aside development tasks I want
to do (both for Commons Imaging and for my own projects), but I think that
it is too much of a handicap not understanding how the underlying system
works.

**Comment by gwlucastrig** [comment]
I'm going to start over on this one.  Please disregard this pull request.

**Pull request closed without merging**